File: 130 Surrounded Regions.py
'''
130.Surrounded Regions
https://leetcode.com/problems/surrounded-regions/

Given an m x n matrix board containing 'X' and 'O',
capture all regions that are 4-directionally surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.

Example 1:

Input: board = [["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]]
Output: [["X","X","X","X"],["X","X","X","X"],["X","X","X","X"],["X","O","X","X"]]
Explanation: Notice that an 'O' should not be flipped if:
- It is on the border, or
- It is adjacent to an 'O' that should not be flipped.
The bottom 'O' is on the border, so it is not flipped.
The other three 'O' form a surrounded region, so they are flipped.

Example 2:

Input: board = [["X"]]
Output: [["X"]]

Constraints:
    m == board.length
    n == board[i].length
    1 <= m, n <= 200
    board[i][j] is 'X' or 'O'
'''
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        rowCount = len(board)
        colCount = len(board[0])

        def mark(sx: int, sy: int):
            if board[sx][sy] != 'O':
                return
            row = [(sx, sy)]
            board[sx][sy] = 'V'
            while row:
                newRow = []
                for x, y in row:
                    for i, j in (x+1, y), (x-1, y), (x, y+1), (x, y-1):
                        if 0 <= i < rowCount and 0 <= j < colCount and board[i][j] == 'O':
                            board[i][j] = 'V'
                            newRow.append((i, j))
                row = newRow

        # 1. From the boundary 'O's, mark all connected 'O' as 'V'
        for r in range(rowCount):
            mark(r, 0)
            mark(r, colCount - 1)
        for c in range(colCount):
            mark(0, c)
            mark(rowCount-1, c)

        # 2. Mark all remaining 'O' to 'X'
        for r in range(rowCount):
            for c in range(colCount):
                if board[r][c] == 'O':
                    board[r][c] = 'X'
        
        # 3. Change back all 'V' to 'O':
        for r in range(rowCount):
            for c in range(colCount):
                if board[r][c] == 'V':
                    board[r][c] = 'O'
        

File: 130 Surrounded Regions.go
/*
130.Surrounded Regions
https://leetcode.com/problems/surrounded-regions/

Given a 2D board containing 'X' and 'O' (the letter O), capture all regions surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.

Example:

X X X X
X O O X
X X O X
X O X X
After running your function, the board should be:

X X X X
X X X X
X X X X
X O X X
Explanation:

Surrounded regions shouldn’t be on the border, which means that any 'O' on the border of the board are not flipped to 'X'.
Any 'O' that is not on the border and it is not connected to an 'O' on the border will be flipped to 'X'.
Two cells are connected if they are adjacent cells connected horizontally or vertically.
*/
var Directions = [8]int{0, 1, 0, -1, 1, 0, -1, 0}

/*
The idea is starting from border, if 'O', apply bfs
to set all 'O' cells connected to boarder as 'V'.
Then go though matrix, update to final result.
This is different approach comparing to 1254. Number of Closed Islands.
*/
func solve(board [][]byte) {
    rowCount := len(board)
    if rowCount == 0 {
        return 
    }
    colCount := len(board[0])
    if colCount == 0 {
        return
    }
    
    bfs := func(i int, j int) {
        stack := [][]int {{i,j}}
        for len(stack) > 0 {
            var newStack [][]int
            for _, p := range stack {
                for d := 0; d < 8; d += 2 {
                    x := p[0] + Directions[d]
                    y := p[1] + Directions[d+1]
                    if x >= 0 && x < rowCount && y >= 0 && y < colCount && board[x][y] == 'O'{
                        // To avoid duplicate entry in stack,
                        // set visited here before pushing into stack
                        board[x][y] = 'V'
                        newStack = append(newStack, []int{x,y})
                    }
                }
            }
            stack = newStack
        }
    }
    
    // 1. BFS from borders
    colBorders := [2]int{0, colCount-1}
    for i := 0; i < rowCount; i++ {
        for _, j := range colBorders {
            if board[i][j] == 'O' {
                board[i][j] = 'V'
                bfs(i, j)
            }
        }
    }
    
    rowBorders := [2]int{0, rowCount-1}
    for j := 0; j < colCount; j++ {
        for _, i := range rowBorders {
            if board[i][j] == 'O' {
                board[i][j] = 'V'
                bfs(i, j)
            }
        }
    }
    
    // 2. Update result
    for i := 0; i < rowCount; i++ {
        for j := 0; j < colCount; j++ {
            if board[i][j] == 'V'{
                board[i][j] = 'O'
            } else if board[i][j] == 'O'{
                board[i][j] = 'X'
            }
        }
    }
}
